sysl2/sysl/ints.go

```diff
@@ -0,0 +1,442 @@
+			isValid = false
+			break
+		}
+	}
+
+	return isValid
+}
+
+func checkDependencies(module *sysl.Module) []*AppDependency {
+	var errStr string
+	deps := []*AppDependency{}
+	errors := []string{}
+	apps := module.GetApps()
+	for appname, app := range module.GetApps() {
+		for epname, endpoint := range app.GetEndpoints() {
+			calls := []*sysl.Call{}
+			makeCalls(endpoint.GetStmt(), calls)
+			for _, call := range calls {
+				targetName := formatAppName(call.GetTarget().GetPart())
+				targetApp := apps[targetName]
+				if targetApp == nil {
+					errStr = fmt.Sprintf("%s <- %s: calls non-existent app %s", appname, epname, targetName)
+					errors = append(errors, errStr)
+				} else {
+					isValid := validatePatterns("abstract", extractApplicationAttr("patterns", targetApp))
+					if !isValid {
+						panic(fmt.Sprintf("call target '%s' must not be ~abstract", targetName))
+					}
+					callEndpoint := call.GetEndpoint()
+					if targetApp.GetEndpoints()[callEndpoint] == nil {
+						errStr = fmt.Sprintf(
+							"%s <- %s: calls non-existent endpoint %s -> %s", appname, epname, targetName, callEndpoint)
+					} else {
+						selfApp := MakeAppElement(appname, epname)
+						targetApp := MakeAppElement(targetName, callEndpoint)
+						dep := MakeAppDependency(selfApp, targetApp)
+						deps = append(deps, dep)
+					}
+				}
+			}
+		}
+	}
+
+	if len(errors) > 0 {
+		panic(fmt.Sprintf("broken deps:\n  %s", strings.Join(errors, "\n")))
+	}
+	return deps
+}
+
+func extractAppNames(dep *AppDependency) []string {
+	apps := []string{}
+	apps = append(apps, dep.Self.Name)
+	apps = append(apps, dep.Target.Name)
+
+	return apps
+}
+
+func extractEndpoints(dep *AppDependency) []string {
+	eps := []string{}
+	eps = append(eps, dep.Self.Endpoint)
+	eps = append(eps, dep.Target.Endpoint)
+
+	return eps
+}
+
+func arrayToMap(a []string) map[string]bool {
+	m := make(map[string]bool)
+	for _, v := range a {
+		m[v] = true
+	}
+
+	return m
+}
+
+func intersection(a, b []string) []string {
+	result := []string{}
+	m := arrayToMap(a)
+	for _, k := range b {
+		if _, ok := m[k]; ok {
+			result = append(result, k)
+		}
+	}
+
+	return result
+}
+
+func subtraction(a, b []string) []string {
+	result := []string{}
+	m := arrayToMap(b)
+	for _, k := range a {
+		if _, ok := m[k]; !ok {
+			result = append(result, k)
+		}
+	}
+
+	return result
+}
+
+func isSub(child, parent []string) bool {
+	isSub := true
+	m := arrayToMap(parent)
+	for _, k := range child {
+		if _, ok := m[k]; !ok {
+			isSub = false
+			break
+		}
+	}
+
+	return isSub
+}
+
+func contains(a string, arr []string) bool {
+	m := arrayToMap(arr)
+
+	return m[a]
+}
+
+func findMatchingApps(module *sysl.Module, excludes []string, integrations []string, deps []*AppDependency) []string {
+	result := []string{}
+	appReStr := fmt.Sprintf(`^(?:%s)(?: *::|$)`, strings.Join(integrations, "|"))
+	re := regexp.MustCompile(appReStr)
+	for _, apps := range deps {
+		appNames := extractAppNames(apps)
```

Comment:
what if combine these two lines to one attaching to AppDependency. Then filter []AppDependency and map to []string.